tests/multigrid/FullFACPreconditioner.cpp

```diff
@@ -0,0 +1,350 @@
+                                             base_level->getBoxes(),
+                                             base_level->getProcessorMapping());
+        Pointer<PatchLevel<NDIM>> dense_level = d_dense_hierarchy->getPatchLevel(ln);
+    }
+}
+
+void
+FullFACPreconditioner::transferToDense(const SAMRAIVectorReal<NDIM, double>& base_x,
+                                       const SAMRAIVectorReal<NDIM, double>& dense_x)
+{
+    Pointer<CartesianGridGeometry<NDIM>> grid_geom = d_hierarchy->getGridGeometry();
+    // We need to create refinement schedules to go from the base to dense hierarchy
+    // TODO: These schedules should be cached and saved between calls to solveSystem.
+    int level_diff = d_dense_hierarchy->getFinestLevelNumber() - d_hierarchy->getFinestLevelNumber();
+    Pointer<RefineAlgorithm<NDIM>> refine_alg = new RefineAlgorithm<NDIM>();
+    // First create the operators needed.
+    for (int comp = 0; comp < dense_x.getNumberOfComponents(); ++comp)
+    {
+        Pointer<RefineOperator<NDIM>> refine_op =
+            grid_geom->lookupRefineOperator(dense_x.getComponentVariable(comp), "CONSERVATIVE_LINEAR_REFINE");
+        int dense_idx = dense_x.getComponentDescriptorIndex(comp);
+        int base_idx = base_x.getComponentDescriptorIndex(comp);
+        refine_alg->registerRefine(dense_idx, base_idx, dense_idx, refine_op);
+    }
+    std::vector<Pointer<RefineSchedule<NDIM>>> refine_scheds(d_hierarchy->getFinestLevelNumber() + 1);
+    // Now create schedules to fill in data.
+    for (int ln = 0; ln <= d_hierarchy->getFinestLevelNumber(); ++ln)
+    {
+        Pointer<PatchLevel<NDIM>> base_level = d_hierarchy->getPatchLevel(ln);
+        Pointer<PatchLevel<NDIM>> dense_level = d_dense_hierarchy->getPatchLevel(ln + level_diff);
+        refine_scheds[ln] = refine_alg->createSchedule(dense_level, base_level);
+        refine_scheds[ln]->fillData(0.0, false);
+    }
+
+    // Now we need to coarsen the data from the dense hierarchy to the dense hierarchy.
+    Pointer<CoarsenAlgorithm<NDIM>> coarsen_alg = new CoarsenAlgorithm<NDIM>();
+    for (int comp = 0; comp < dense_x.getNumberOfComponents(); ++comp)
+    {
+        Pointer<CoarsenOperator<NDIM>> coarsen_op =
+            grid_geom->lookupCoarsenOperator(dense_x.getComponentVariable(comp), "CUBIC_COARSEN");
+        int dense_idx = dense_x.getComponentDescriptorIndex(comp);
+        coarsen_alg->registerCoarsen(dense_idx, dense_idx, coarsen_op);
+    }
+    std::vector<Pointer<CoarsenSchedule<NDIM>>> coarsen_scheds(level_diff);
+    for (int ln = level_diff - 1; ln >= 0; --ln)
+    {
+        Pointer<PatchLevel<NDIM>> dst_level = d_dense_hierarchy->getPatchLevel(ln);
+        Pointer<PatchLevel<NDIM>> src_level = d_dense_hierarchy->getPatchLevel(ln + 1);
+        coarsen_scheds[ln] = coarsen_alg->createSchedule(dst_level, src_level);
+        coarsen_scheds[ln]->coarsenData();
+    }
```

Comment:
This isn't strictly necessary. For multigrid, we only need to copy the data down to the first level that covers the patch hierarchy. This is done in the "refinement" operations above.

Reply:
I'm leaving this for now, it shouldn't (and doesn't in the test) affect the solution.